test-crypt-mt.c

```diff
@@ -0,0 +1,189 @@
+#elif INCLUDE_gost_yescrypt
+#define HASH_PREFIX "$gy$"
+#elif INCLUDE_bcrypt
+#define HASH_PREFIX "$2b$"
+#elif INCLUDE_bcrypt_y
+#define HASH_PREFIX "$2y$"
+#elif INCLUDE_bcrypt_a
+#define HASH_PREFIX "$2a$"
+#elif INCLUDE_sha512crypt
+#define HASH_PREFIX "$6$"
+#elif INCLUDE_sha256crypt
+#define HASH_PREFIX "$5$"
+#elif INCLUDE_sha1crypt
+#define HASH_PREFIX "$sha1"
+#elif INCLUDE_sunmd5
+#define HASH_PREFIX "$md5"
+#elif INCLUDE_md5crypt
+#define HASH_PREFIX "$1$"
+#elif INCLUDE_bsdicrypt
+#define HASH_PREFIX "_"
+#elif INCLUDE_bigcrypt || INCLUDE_descrypt
+#define HASH_PREFIX ""
+#endif
+
+#if defined HAVE_PTHREAD && defined HASH_PREFIX
+
+/* for determinism, we don't use auto-entropy either.  */
+static const char rbytes1[] =
+  "opm+IVsGxcmb73BXTSjkWueVJMx1W1KAIV0lPQctV2Hxc7Nc3UCoi1jN3nW6UlFZ";
+static const char rbytes2[] =
+  "MHTAZtYAvBI54WLE2vp+ekStQhfp0uakGbX397u/DvffB/hvb/ry95MWOKWQlu5A";
+
+static const char pw1[] = "fraggle";
+static const char pw2[] = "doozer";
+
+static void *
+child (void *ARG_UNUSED (data))
+{
+  char *setting, *hash;
+
+  setting = crypt_gensalt (HASH_PREFIX, 0, rbytes2, 64);
+  if (!setting)
+    {
+      printf ("ERROR: child: crypt_gensalt failed: %s\n", strerror (errno));
+      return ((void *)1);
+    }
+  else
+    printf ("ok: child: crypt_gensalt = %s\n", setting);
+
+  hash = crypt (pw2, setting);
+  if (!hash)
+    {
+      printf ("ERROR: child: crypt failed: %s\n", strerror (errno));
+      return ((void *)1);
+    }
+  else if (hash[0] == '*' || hash[0] == '\0')
+    {
+      printf ("ERROR: child: crypt failed (\"%s\"): %s\n",
+              hash, strerror (errno));
+      return ((void *)1);
+    }
+  else
+    printf ("ok: child: crypt = %s\n", hash);
+  return 0;
+}
+
+int main(void)
+{
+  char save_setting[CRYPT_GENSALT_OUTPUT_SIZE];
+  char save_hash[CRYPT_OUTPUT_SIZE];
+  char *setting, *hash;
+  pthread_t th;
+  int err;
+  void *status;
+
+  if (setvbuf (stdout, 0, _IOLBF, 0))
+    {
+      printf ("ERROR: setvbuf: %s\n", strerror (errno));
+      return 0;
+    }
+
+  setting = crypt_gensalt (HASH_PREFIX, 0, rbytes1, 64);
+  if (!setting)
+    {
+      printf ("ERROR: parent: crypt_gensalt failed: %s\n", strerror (errno));
+      return 1;
+    }
+  else
+    printf ("ok: parent: crypt_gensalt = %s\n", setting);
+  if (strlen (setting) + 1 > sizeof save_setting)
+    {
+      printf ("ERROR: crypt_gensalt output is too long (%zu > %zu)\n",
+              strlen (setting) + 1, sizeof save_setting);
+      return 1;
+    }
+  strcpy (save_setting, setting);
+
+  hash = crypt (pw1, setting);
+  if (!hash)
+    {
+      printf ("ERROR: parent: crypt failed: %s\n", strerror (errno));
+      return 1;
+    }
+  else if (hash[0] == '*' || hash[0] == '\0')
+    {
+      printf ("ERROR: parent: crypt failed (\"%s\"): %s\n",
+              hash, strerror (errno));
+      return 1;
+    }
+  else
+    printf ("ok: parent: crypt = %s\n", hash);
+  if (strlen (hash) + 1 > sizeof save_hash)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/manual/besser82/libxcrypt/pullRequest?prid=5024327)